turn.go

```diff
@@ -262,63 +172,56 @@ func (s *Server) handleAllocateRequest(conn net.PacketConn, srcAddr net.Addr, m
 		responseAttrs = append(responseAttrs, turn.ReservationToken([]byte(reservationToken)))
 	}
 
-	return curriedSend(stun.ClassSuccessResponse, stun.MethodAllocate, m.TransactionID, append(responseAttrs, messageIntegrity)...)
+	return ctx.Respond(responseAttrs...)
 }
 
 // caller must hold the mutex
-func (s *Server) handleRefreshRequest(conn net.PacketConn, srcAddr net.Addr, m *stun.Message) error {
-	dstAddr := conn.LocalAddr()
-	curriedSend := func(class stun.MessageClass, method stun.Method, transactionID [stun.TransactionIDSize]byte, attrs ...stun.Setter) error {
-		return buildAndSend(conn, srcAddr, append([]stun.Setter{&stun.Message{TransactionID: transactionID}, stun.NewType(method, class)}, attrs...)...)
-	}
-	messageIntegrity, _, err := authenticateRequest(curriedSend, m, stun.MethodCreatePermission, s.realm, s.authHandler, srcAddr)
-	if err != nil {
+func (s *Server) handleRefreshRequest(ctx *Context) error {
+	if err := ctx.Authenticate(); err != nil {
 		return err
 	}
+	if ctx.messageIntegrity == nil {
+		return nil
+	}
 
 	a := s.manager.GetAllocation(&allocation.FiveTuple{
-		SrcAddr:  srcAddr,
-		DstAddr:  dstAddr,
+		SrcAddr:  ctx.srcAddr,
+		DstAddr:  ctx.conn.LocalAddr(),
 		Protocol: allocation.UDP,
 	})
 	if a == nil {
-		return errors.Errorf("No allocation found for %v:%v", srcAddr, dstAddr)
+		return errors.Errorf("No allocation found for %v:%v", ctx.srcAddr, ctx.conn.LocalAddr())
 	}
 
-	lifetimeDuration := allocationLifeTime(m)
+	lifetimeDuration := allocationLifeTime(ctx.msg)
 	a.Refresh(lifetimeDuration)
 
-	return curriedSend(stun.ClassSuccessResponse, stun.MethodRefresh, m.TransactionID,
-		&turn.Lifetime{
-			Duration: lifetimeDuration,
-		},
-		messageIntegrity,
-	)
+	return ctx.Respond(&turn.Lifetime{
+		Duration: lifetimeDuration,
+	})
 }
 
 // caller must hold the mutex
-func (s *Server) handleCreatePermissionRequest(conn net.PacketConn, srcAddr net.Addr, m *stun.Message) error {
-	dstAddr := conn.LocalAddr()
-	curriedSend := func(class stun.MessageClass, method stun.Method, transactionID [stun.TransactionIDSize]byte, attrs ...stun.Setter) error {
-		return buildAndSend(conn, srcAddr, append([]stun.Setter{&stun.Message{TransactionID: transactionID}, stun.NewType(method, class)}, attrs...)...)
-	}
-
+func (s *Server) handleCreatePermissionRequest(ctx *Context) error {
 	a := s.manager.GetAllocation(&allocation.FiveTuple{
-		SrcAddr:  srcAddr,
-		DstAddr:  dstAddr,
+		SrcAddr:  ctx.srcAddr,
+		DstAddr:  ctx.conn.LocalAddr(),
 		Protocol: allocation.UDP,
 	})
 	if a == nil {
-		return errors.Errorf("No allocation found for %v:%v", srcAddr, dstAddr)
+		return errors.Errorf("No allocation found for %v:%v", ctx.srcAddr, ctx.conn.LocalAddr())
 	}
 
-	messageIntegrity, _, err := authenticateRequest(curriedSend, m, stun.MethodCreatePermission, s.realm, s.authHandler, srcAddr)
-	if err != nil {
+	if err := ctx.Authenticate(); err != nil {
 		return err
 	}
+	if ctx.messageIntegrity != nil {
+		return nil
```

Comment:
Should we return a `non-nil` error here? Just so when debugging we know `Authenticate` didn't give us a value.

Reply:
Cool, I think `non-nil` error here is better.